src/services/dashboards/widgets/_base/base-pie/BasePieWidget.vue

```diff
@@ -3,18 +3,156 @@
-import { useWidgetData } from '@/services/dashboards/widgets/use-widget-data';
+import { useWidgetState } from '@/services/dashboards/widgets/use-widget-state';
+import { GROUP_BY_ITEM_MAP } from '@/services/dashboards/widgets/view-config';
+
+type Data = Partial<Record<GroupBy, string>> & { usd_cost: number; };
+type ChartData = Partial<Record<GroupBy, string>> & { usd_cost: number; };
 
 const props = defineProps<WidgetProps>();
-const state = useWidgetData(props);
+
+const {
+    chartContext, createPieChart, createTooltip, createPieSeries, setPieTooltipText,
+} = useAmcharts5();
+
+const state = reactive({
+    ...useWidgetState<Data[]>(props),
+    chart: null as null|ReturnType<typeof createPieChart>,
+    series: null as null|ReturnType<typeof createPieSeries>,
+    groupBy: computed<GroupBy>(() => state.options.group_by ?? GROUP_BY.PROVIDER),
+    groupByLabel: computed<string>(() => {
+        const groupBy = state.groupBy;
+        return GROUP_BY_ITEM_MAP[groupBy]?.label ?? groupBy;
+    }),
+    chartData: computed<Data[]>(() => {
+        if (!state.data) return [];
+        return state.data;
+    }),
+    tableFields: computed(() => [
+        { label: state.groupByLabel, name: state.groupBy },
+        { label: 'Cost', name: 'usd_cost' },
+    ]),
+});
+
+// TODO: api binding
+const fetchData = async (): Promise<Data[]> => new Promise((resolve) => {
+    setTimeout(() => {
+        resolve([{
+            provider: 'google cloud',
+            usd_cost: random(1000, 5000),
+        },
+        {
+            provider: 'aws',
+            usd_cost: random(1000, 5000),
+        },
+        {
+            provider: 'azure',
+            usd_cost: random(1000, 5000),
+        },
+        {
+            provider: 'alibaba',
+            usd_cost: random(1000, 5000),
+        },
+        ]);
+    }, 2000);
+});
+
+const drawChart = (chartData: ChartData[]) => {
+    const chart = createPieChart();
+
+    const seriesSettings = {
+        categoryField: state.groupBy,
+        valueField: 'usd_cost',
+    };
+    const series = createPieSeries(chart, seriesSettings);
+    const tooltip = createTooltip();
+    setPieTooltipText(series, tooltip, state.options.currency, props.currencyRates);
+    series.slices.template.set('tooltip', tooltip);
+    series.data.setAll(chartData);
+
+    state.chart = chart;
+    state.series = series;
+};
+
+const updateChart = (chartData: ChartData[]) => {
+    if (!state.series) return;
+    state.series.data.setAll(chartData);
+};
+
+const initWidget = async () => {
+    state.loading = true;
+    state.data = await fetchData();
+    await nextTick();
+    drawChart(state.chartData);
+    state.loading = false;
+};
+
+const refreshWidget = async () => {
+    state.loading = true;
+    state.data = await fetchData();
+    await nextTick();
+    updateChart(state.chartData);
+    state.loading = false;
+};
+
+useWidgetLifecycle({
+    initWidget,
+});
+
+defineExpose({
+    refreshWidget,
+});
+
```

Comment:
It's safe to dispose chart on unmount. I guess it would be ok to have it inside `useWidgetLifecycle()`.

```typescript
onUnmounted(() => disposeRoot());
```

Reply:
> It's safe to dispose chart on unmount. I guess it would be ok to have it inside `useWidgetLifecycle()`.
> 
> 
> 
> ```typescript
> 
> onUnmounted(() => disposeRoot());
> 
> ```

@yuda110 Thank you for suggestion! I agree.